README.md
# LUA LiPo Watcher
Simple battery monitoring with total flight-pack voltage or Rx-voltage

Works in DC/DS-14/16/24 with frimware 4.11 and up

Current info available on https://www.rc-thoughts.com/lipo-watcher/

Tero from RC-Thoughts 2019

![Display-picture](https://www.rc-thoughts.com/wp-content/uploads/2019/10/RCT-LiPoWatcher_13.gif)


CHANGELOG.txt
------------------------------------------------------------
                   Lua LiPo Watcher
------------------------------------------------------------

 Tero Salminen RC-Thoughts.com (c) 2017 www.rc-thoughts.com

------------------------------------------------------------

2019-10-05  Tero Salminen  <tero@rc-thoughts.com>

    * v.1.3 Added clock-display
            Added color for 12/24
            Separate version for 14/16 and 12/24
            Redefined bar-calculation

2017-10-28  Tero Salminen  <tero@rc-thoughts.com>

    * v.1.2 Changed the way calculation is done, changed to 
            max. voltage tracking, tested in DS-16

2017-10-22  Tero Salminen  <tero@rc-thoughts.com>

    * v.1.1 Added possibility to adjust averaging value

2017-10-22  Tero Salminen  <tero@rc-thoughts.com>

    * v.1.0 First release


Source/RCT-Li16.lua
--[[
    ---------------------------------------------------------
    LiPo Watcher takes flight pack voltage and calculates
    cell-voltage from that. Cell-voltage is then used to
    determine LiPo-charge. 
    
    Voltage is moothed heavily to eliminate spikes generating
    false alarms.
    
    User has a possibility to set an alarm with desired point
    with audio-file if desired.
    
    Requires DC/DS-14/16 with firmware 4.22 or up.
    ---------------------------------------------------------
    LiPo Watcher is part of RC-Thoughts Jeti Tools.
    ---------------------------------------------------------
            Released under MIT-license by 
        Tero Salminen @ RC-Thoughts.com 2017-2019
    ---------------------------------------------------------
--]]
collectgarbage()
----------------------------------------------------------------------
-- Locals for the application
local roll,voltTot,cellVolt,cellPerc,playDone=0,0,0,-1,false
local sensId,sensPa,cellCnt,alarmVal,alarmFile=0,0,0,0,false
local timeNow,timeLast,timeFill,voltMax,rX=0,0,0,0
----------------------------------------------------------------------
-- Read translations
local function setLanguage()
    local lng=system.getLocale()
    local file=io.readall("Apps/Lang/RCT-LiP2.jsn")
    local obj=json.decode(file)
    if(obj) then
        trans19=obj[lng] or obj[obj.default]
    end
    collectgarbage()
end
----------------------------------------------------------------------
local function LiPoGauge(percentage,ox,oy)
    -- Fuel bar for Big Screen
    lcd.drawRectangle (8+ox,53+oy,30,11)
    lcd.drawRectangle (8+ox,41+oy,30,11)  
    lcd.drawRectangle (8+ox,29+oy,30,11)  
    lcd.drawRectangle (8+ox,17+oy,30,11)  
    lcd.drawRectangle (8+ox,5+oy,30,11)
    -- Bar chart
    if(cellPerc >= 0) then
        local nSolidBar=math.floor(cellPerc / 20)
        local nFracBar=(cellPerc-nSolidBar * 20) / 20
        local i
        -- Solid bars
        for i=0,nSolidBar-1,1 do 
            lcd.drawFilledRectangle (9+ox,54-i*12+oy,28,9) 
        end  
        -- Fractional bar
        local y=math.ceil(54-nSolidBar*12+(1-nFracBar)*9)
        lcd.drawFilledRectangle (9+ox,y+oy,28,9*nFracBar)
        lcd.setColor(txtr,txtg,txtb)
    end
    collectgarbage()
end
----------------------------------------------------------------------
-- Telemetry for Small-display
local function dispLiPoSmall(percentage,ox,oy)
    -- Fuel bar 
    lcd.drawRectangle (71+ox,4+oy,16,15)
    lcd.drawRectangle (54+ox,4+oy,16,15)  
    lcd.drawRectangle (37+ox,4+oy,16,15)  
    lcd.drawRectangle (20+ox,4+oy,16,15)  
    lcd.drawRectangle (3+ox,4+oy,16,15)
    -- Bar chart
    if(cellPerc >= 0) then
        local nSolidBar=math.floor(cellPerc / 20)
        local nFracBar=(cellPerc-nSolidBar * 20) / 20
        local i
        -- Solid bars
        for i=0,nSolidBar-1,1 do 
            lcd.drawFilledRectangle (4+i*17+ox,5+oy,14,13) 
        end  
        -- Fractional bar
        local x=math.floor(4+nSolidBar*17)
        lcd.drawFilledRectangle (x+ox,5+oy,15*nFracBar,13)
    end
end
----------------------------------------------------------------------
-- Telemetry for Clock-display
local function dispLiPoClock(percentage,ox,oy)
    -- Fuel bar 
    lcd.drawRectangle (59+ox,3+oy,13,13)
    lcd.drawRectangle (45+ox,3+oy,13,13)  
    lcd.drawRectangle (31+ox,3+oy,13,13)  
    lcd.drawRectangle (17+ox,3+oy,13,13)  
    lcd.drawRectangle (3+ox,3+oy,13,13)
    -- Bar chart
    if(cellPerc >= 0) then
        local nSolidBar=math.floor(cellPerc / 20)
        local nFracBar=(cellPerc-nSolidBar * 20) / 20
        local i
        -- Solid bars
        for i=0,nSolidBar-1,1 do 
            lcd.drawFilledRectangle (4+i*14+ox,4+oy,11,11) 
        end  
        -- Fractional bar
        local x=math.floor(4+nSolidBar*14)
        lcd.drawFilledRectangle (x+ox,4+oy,11*nFracBar,11)
    end
end
----------------------------------------------------------------------
-- Draw the telemetry windows
local function dispLiPo(width,height)
    if(height==69)then -- Big window
        lcd.drawText(140-lcd.getTextWidth(FONT_MINI,string.format(trans19.winLabel)),3,string.format(trans19.winLabel),FONT_MINI)
        if (cellPerc==-1) then
            lcd.drawText(140-lcd.getTextWidth(FONT_MAXI,"-%"),16,"-%",FONT_MAXI)
            lcd.drawText(140-lcd.getTextWidth(FONT_MINI,"RC-Thoughts.com"),53,"RC-Thoughts.com",FONT_MINI)
            else
            lcd.drawText(140-lcd.getTextWidth(FONT_MAXI,string.format("%s%%",cellPerc)),16,string.format("%s%%",cellPerc),FONT_MAXI)
            lcd.drawText(140-lcd.getTextWidth(FONT_MINI,string.format("%s %.2fV",trans19.lipoLabel,cellVolt)),53,string.format("%s %.2fV",trans19.lipoLabel,cellVolt),FONT_MINI)
        end
        LiPoGauge(cellPerc,1,0)
        elseif(width==75 and height==30) then -- Clock Display
                if (cellPerc==-1) then
                    lcd.drawText(40-lcd.getTextWidth(FONT_MINI,"-%"),12,"-%",FONT_BOLD)
                else
                    lcd.drawText(40-lcd.getTextWidth(FONT_BOLD,string.format("%s%%",cellPerc)),12,string.format("%s%%",cellPerc),FONT_BOLD)
                end
                lcd.drawText(72-lcd.getTextWidth(FONT_MINI,string.format("%.2fV",cellVolt)),16,string.format("%.2fV",cellVolt),FONT_MINI)
                dispLiPoClock(cellPerc,1,0)
    else -- Small window
        if (cellPerc==-1) then
            lcd.drawText(145-lcd.getTextWidth(FONT_BIG,"-%"),1,"-%",FONT_BIG)
            else
            lcd.drawText(145-lcd.getTextWidth(FONT_BIG,string.format("%s%%",cellPerc)),1,string.format("%s%%",cellPerc),FONT_BIG)
        end
        dispLiPoSmall(cellPerc,1,0)
    end
    collectgarbage()
end
----------------------------------------------------------------------
-- Store settings when changed by user
local function sensorChanged(value)
    sensId=sensorsAvailable[value].id
    sensPa=sensorsAvailable[value].param
    system.pSave("sensId",sensId)
    system.pSave("sensPa",sensPa)
end

local function cellCntChanged(value)
    cellCnt=value
    system.pSave("cellCnt",cellCnt)
end

local function alarmValChanged(value)
    alarmVal=value
    system.pSave("alarmVal",alarmVal)
end

local function alarmFileChanged(value)
    alarmFile=value
    system.pSave("alarmFile",alarmFile)
end
----------------------------------------------------------------------
-- Draw the main form (Application inteface)
local function initForm(subform)
    -- List sensors only if menu is active to preserve memory at runtime 
    -- (measured up to 25% save if menu is not opened)
    sensorsAvailable = {}
    local sensors = system.getSensors()
    local sensorTx = system.getTxTelemetry()
    local list={}
    local curIndex1 = -1
    local descr = ""
    -- Add some of RX Telemetry items to beginning in list of sensors, get name from translation
    list[#list + 1] = string.format("%s",trans19.sensorRx1)
    sensorsAvailable[#sensorsAvailable + 1] = {["unit"] = "V", ["param"] = 1,["id"] = 999,["label"] = trans19.sensorRx1}
    if(sensId == 999 and sensPa == 1) then
        curIndex1 = 1
    end
    list[#list + 1] = string.format("%s",trans19.sensorRx2)
    sensorsAvailable[#sensorsAvailable + 1] = {["unit"] = "V", ["param"] = 2,["id"] = 999,["label"] = trans19.sensorRx2}
    if(sensId == 999 and sensPa == 2) then
        curIndex1 = 2
    end
    list[#list + 1] = string.format("%s",trans19.sensorRxB)
    sensorsAvailable[#sensorsAvailable + 1] = {["unit"] = "V", ["param"] = 3,["id"] = 999,["label"] = trans19.sensorRxB}
    if(sensId == 999 and sensPa == 3) then
        curIndex1 = 3
    end
    -- Create a list of normal sensors
    for index,sensor in ipairs(sensors) do 
        if(sensor.param == 0) then
            descr = sensor.label
            else
            list[#list + 1] = string.format("%s - %s",descr,sensor.label)
            sensorsAvailable[#sensorsAvailable + 1] = sensor
            if(sensor.id == sensId and sensor.param == sensPa) then
                curIndex1 =# sensorsAvailable
            end
        end
    end
    collectgarbage()
    
    local form,addRow,addLabel=form,form.addRow,form.addLabel
    local addIntbox,addSelectbox=form.addIntbox,form.addSelectbox
    local addInputbox,addCheckbox=form.addInputbox,form.addCheckbox
    local addAudioFilebox,setButton=form.addAudioFilebox,form.setButton
    local addTextbox=form.addTextbox
    
    addRow(1)
    addLabel({label="---     RC-Thoughts Jeti Tools      ---",font=FONT_BIG})
    
    addRow(1)
    addLabel({label=trans19.labelSensor,font=FONT_BOLD})
    
    addRow(2)
    addLabel({label=trans19.sensorSel})
    --addSelectbox(sensList,curIndex,true,sensorChanged)
    addSelectbox(list,curIndex1,true,sensorChanged)
    
    addRow(2)
    addLabel({label=trans19.cellCount,width=220})
    addIntbox(cellCnt,0,24,0,0,1,cellCntChanged)
    
    addRow(1)
    addLabel({label=trans19.labelAlarm,font=FONT_BOLD})
    
    addRow(2)
    addLabel({label=trans19.alarmValue,width=220})
    addIntbox(alarmVal,0,99,0,0,1,alarmValChanged)
    
    form.addRow(2)
    addLabel({label=trans19.voiceFile})
    addAudioFilebox(alarmFile,alarmFileChanged)
    
    addRow(1)
    addLabel({label="Powered by RC-Thoughts.com-"..lipoVersion.." ",font=FONT_MINI,alignRight=true})
    
    formID=1
    collectgarbage()
end
----------------------------------------------------------------------
-- Table for binding cell-voltage to percentage
local percentList={{3,0},{3.093,1},{3.196,2},{3.301,3},{3.401,4},{3.477,5},{3.544,6},{3.601,7},{3.637,8},{3.664,9},{3.679,10},{3.683,11},{3.689,12},{3.692,13},{3.705,14},{3.71,15},{3.713,16},{3.715,17},{3.72,18},{3.731,19},{3.735,20},{3.744,21},{3.753,22},{3.756,23},{3.758,24},{3.762,25},{3.767,26},{3.774,27},{3.78,28},{3.783,29},{3.786,30},{3.789,31},{3.794,32},{3.797,33},{3.8,34},{3.802,35},{3.805,36},{3.808,37},{3.811,38},{3.815,39},{3.818,40},{3.822,41},{3.825,42},{3.829,43},{3.833,44},{3.836,45},{3.84,46},{3.843,47},{3.847,48},{3.85,49},{3.854,50},{3.857,51},{3.86,52},{3.863,53},{3.866,54},{3.87,55},{3.874,56},{3.879,57},{3.888,58},{3.893,59},{3.897,60},{3.902,61},{3.906,62},{3.911,63},{3.918,64},{3.923,65},{3.928,66},{3.939,67},{3.943,68},{3.949,69},{3.955,70},{3.961,71},{3.968,72},{3.974,73},{3.981,74},{3.987,75},{3.994,76},{4.001,77},{4.007,78},{4.014,79},{4.021,80},{4.029,81},{4.036,82},{4.044,83},{4.052,84},{4.062,85},{4.074,86},{4.085,87},{4.095,88},{4.105,89},{4.111,90},{4.116,91},{4.12,92},{4.125,93},{4.129,94},{4.135,95},{4.145,96},{4.176,97},{4.179,98},{4.193,99},{4.2,100}}
----------------------------------------------------------------------
-- Count percentage from cell voltage
function percCell(cellVoltage)
    result=0
    if(cellVoltage > 4.2 or cellVoltage < 3.00)then
        if(cellVoltage > 4.2)then
            result=100
        end
        if(cellVoltage < 3.00)then
            result=0
        end
        else
        for i,v in ipairs(percentList) do
            if(v[1] >= cellVoltage)then
                result=v[2]
                break
            end
        end
    end
    collectgarbage()
    return result
end
----------------------------------------------------------------------
-- Create a rolling average of voltage to prevent false alarms
function rollingAverage(period)
    local t={}
    function sum(a,...)
        if a then return a+sum(...) else return 0 end
    end
    function average(n)
        if #t==period then table.remove(t,1) end
        t[#t+1]=n
        return sum(table.unpack(t)) / #t
    end
    collectgarbage()
    return average
end
----------------------------------------------------------------------
-- Runtime functions,read sensor,convert to percentage
local function loop()
    local sensor = {}
    local sensorRx = system.getTxTelemetry()
    if(sensId == 999) then
        if(sensPa == 1) then
            sensor.value = sensorRx.rx1Voltage
            elseif (sensPa == 2) then
            sensor.value = sensorRx.rx2Voltage
            elseif (sensPa == 3) then
            sensor.value = sensorRx.rxBVoltage
        end
        if(sensorRx.rx1Percent > 1)then
            sensor.valid = true
        else
            sensor.valid = false
        end
        else
        sensor = system.getSensorByID(sensId,sensPa)
    end

    local timeNow = system.getTimeCounter()

    if(sensor and sensor.valid) then
        voltTot=sensor.value
        -- Fill table at start
        if(timeFill == 0) then
            timeFill = timeNow + 2000
        end
        if(timeNow <= timeFill) then
            cellVolt = (roll(voltTot)/cellCnt)
        end
        -- 
        if(timeNow >= timeLast + 1000) then
            cellVolt = (roll(voltTot)/cellCnt)
            timeLast = timeNow
            voltMax = 0
        end
        -- Track maximum voltage seen during this collection cycle
		if voltTot > voltMax then	
			voltMax = voltTot
		end
        
        -- Calculate cell-percentage from LiPo-table
        cellPerc=percCell(cellVolt)
        -- Take care of alarm if alarm is enabled, make sure it plays only once
        if(not playDone and alarmVal > 0 and cellPerc <= alarmVal and alarmFile ~= "...") then
            system.playFile(alarmFile,AUDIO_QUEUE)
            system.playNumber(cellPerc,0,"%")
            playDone=true   
        end
        else
        -- If we have no sensor set cell-percentage to -1 for screen-identification
        cellPerc = -1
        timeFill = 0
    end   
    -- If percentage is above alarm-level enable alarm
    if(cellPerc > alarmVal) then
        playDone=false
    end
    collectgarbage()
end
----------------------------------------------------------------------
-- Application initialization
local function init()
    local pLoad,registerForm,registerTelemetry=system.pLoad,system.registerForm,system.registerTelemetry
    sensId=pLoad("sensId",0)
    sensPa=pLoad("sensPa",0)
    cellCnt=pLoad("cellCnt",0)
    alarmVal=pLoad("alarmVal",0)
    alarmFile=pLoad("alarmFile","...")
    registerForm(1,MENU_APPS,trans19.appName,initForm)
    registerTelemetry(1,trans19.appName,0,dispLiPo)
    -- Set average-calculation
    roll = rollingAverage(10)
    collectgarbage()
end
----------------------------------------------------------------------
lipoVersion="v.1.3"
setLanguage()
collectgarbage()
return {init=init,loop=loop,author="RC-Thoughts",version=lipoVersion,name=trans19.appName}


Source/RCT-Li24.lua
--[[
    ---------------------------------------------------------
    LiPo Watcher takes flight pack voltage and calculates
    cell-voltage from that. Cell-voltage is then used to
    determine LiPo-charge. 
    
    Voltage is moothed heavily to eliminate spikes generating
    false alarms.
    
    User has a possibility to set an alarm with desired point
    with audio-file if desired.
    
    Requires DC/DS-14/16 with firmware 4.22 or up.
    ---------------------------------------------------------
    LiPo Watcher is part of RC-Thoughts Jeti Tools.
    ---------------------------------------------------------
            Released under MIT-license by 
        Tero Salminen @ RC-Thoughts.com 2017-2019
    ---------------------------------------------------------
--]]
collectgarbage()
----------------------------------------------------------------------
-- Locals for the application
local roll,voltTot,cellVolt,cellPerc,playDone=0,0,0,-1,false
local sensId,sensPa,cellCnt,alarmVal,alarmFile=0,0,0,0,false
local timeNow,timeLast,timeFill,voltMax,rX=0,0,0,0
----------------------------------------------------------------------
-- Read translations
local function setLanguage()
    local lng=system.getLocale()
    local file=io.readall("Apps/Lang/RCT-Li24.jsn")
    local obj=json.decode(file)
    if(obj) then
        trans19=obj[lng] or obj[obj.default]
    end
    collectgarbage()
end
----------------------------------------------------------------------
local function LiPoGauge(percentage,ox,oy)
    local txtr,txtg,txtb
    local bgr,bgg,bgb = lcd.getBgColor()
    if (bgr+bgg+bgb)/3 >128 then
        txtr,txtg,txtb = 0,0,0
        else
        txtr,txtg,txtb = 255,255,255
    end
    -- Fuel bar for Big Screen
    lcd.drawRectangle (8+ox,53+oy,30,11)
    lcd.drawRectangle (8+ox,41+oy,30,11)  
    lcd.drawRectangle (8+ox,29+oy,30,11)  
    lcd.drawRectangle (8+ox,17+oy,30,11)  
    lcd.drawRectangle (8+ox,5+oy,30,11)
    -- Bar chart
    if(cellPerc >= 0) then
        if(cellPerc <= alarmVal) then
            lcd.setColor(220,0,0) -- Red
        else
            lcd.setColor(0,220,0) -- Green
        end
        local nSolidBar=math.floor(cellPerc / 20)
        local nFracBar=(cellPerc-nSolidBar * 20) / 20
        local i
        -- Solid bars
        for i=0,nSolidBar-1,1 do 
            lcd.drawFilledRectangle (9+ox,54-i*12+oy,28,9) 
        end  
        -- Fractional bar
        local y=math.ceil(54-nSolidBar*12+(1-nFracBar)*9)
        lcd.drawFilledRectangle (9+ox,y+oy,28,9*nFracBar)
        lcd.setColor(txtr,txtg,txtb)
    end
    collectgarbage()
end
----------------------------------------------------------------------
-- Telemetry for Small-display
local function dispLiPoSmall(percentage,ox,oy)
    local txtr,txtg,txtb
    local bgr,bgg,bgb = lcd.getBgColor()
    if (bgr+bgg+bgb)/3 >128 then
        txtr,txtg,txtb = 0,0,0
        else
        txtr,txtg,txtb = 255,255,255
    end
    -- Fuel bar 
    lcd.drawRectangle (71+ox,4+oy,16,15)
    lcd.drawRectangle (54+ox,4+oy,16,15)  
    lcd.drawRectangle (37+ox,4+oy,16,15)  
    lcd.drawRectangle (20+ox,4+oy,16,15)  
    lcd.drawRectangle (3+ox,4+oy,16,15)
    -- Bar chart
    if(cellPerc >= 0) then
        if(cellPerc <= alarmVal) then
            lcd.setColor(220,0,0) -- Red
        else
            lcd.setColor(0,220,0) -- Green
        end
        local nSolidBar=math.floor(cellPerc / 20)
        local nFracBar=(cellPerc-nSolidBar * 20) / 20
        local i
        -- Solid bars
        for i=0,nSolidBar-1,1 do 
            lcd.drawFilledRectangle (4+i*17+ox,5+oy,14,13) 
        end  
        -- Fractional bar
        local x=math.floor(4+nSolidBar*17)
        lcd.drawFilledRectangle (x+ox,5+oy,15*nFracBar,13)
    end
end
----------------------------------------------------------------------
-- Telemetry for Clock-display
local function dispLiPoClock(percentage,ox,oy)
    local txtr,txtg,txtb
    local bgr,bgg,bgb = lcd.getBgColor()
    if (bgr+bgg+bgb)/3 >128 then
        txtr,txtg,txtb = 0,0,0
        else
        txtr,txtg,txtb = 255,255,255
    end
    -- Fuel bar 
    lcd.drawRectangle (59+ox,3+oy,13,13)
    lcd.drawRectangle (45+ox,3+oy,13,13)  
    lcd.drawRectangle (31+ox,3+oy,13,13)  
    lcd.drawRectangle (17+ox,3+oy,13,13)  
    lcd.drawRectangle (3+ox,3+oy,13,13)
    -- Bar chart
    if(cellPerc >= 0) then
        if(cellPerc <= alarmVal) then
            lcd.setColor(220,0,0) -- Red
        else
            lcd.setColor(0,220,0) -- Green
        end
        local nSolidBar=math.floor(cellPerc / 20)
        local nFracBar=(cellPerc-nSolidBar * 20) / 20
        local i
        -- Solid bars
        for i=0,nSolidBar-1,1 do 
            lcd.drawFilledRectangle (4+i*14+ox,4+oy,11,11) 
        end  
        -- Fractional bar
        local x=math.floor(4+nSolidBar*14)
        lcd.drawFilledRectangle (x+ox,4+oy,11*nFracBar,11)
    end
end
----------------------------------------------------------------------
-- Draw the telemetry windows
local function dispLiPo(width,height)
    if(height==69)then -- Big window
        lcd.drawText(140-lcd.getTextWidth(FONT_MINI,string.format(trans19.winLabel)),3,string.format(trans19.winLabel),FONT_MINI)
        if (cellPerc==-1) then
            lcd.drawText(140-lcd.getTextWidth(FONT_MAXI,"-%"),16,"-%",FONT_MAXI)
            lcd.drawText(140-lcd.getTextWidth(FONT_MINI,"RC-Thoughts.com"),53,"RC-Thoughts.com",FONT_MINI)
            else
            lcd.drawText(140-lcd.getTextWidth(FONT_MAXI,string.format("%s%%",cellPerc)),16,string.format("%s%%",cellPerc),FONT_MAXI)
            lcd.drawText(140-lcd.getTextWidth(FONT_MINI,string.format("%s %.2fV",trans19.lipoLabel,cellVolt)),53,string.format("%s %.2fV",trans19.lipoLabel,cellVolt),FONT_MINI)
        end
        LiPoGauge(cellPerc,1,0)
        elseif(width==75 and height==30) then -- Clock Display
                if (cellPerc==-1) then
                    lcd.drawText(40-lcd.getTextWidth(FONT_MINI,"-%"),12,"-%",FONT_BOLD)
                else
                    lcd.drawText(40-lcd.getTextWidth(FONT_BOLD,string.format("%s%%",cellPerc)),12,string.format("%s%%",cellPerc),FONT_BOLD)
                end
                lcd.drawText(72-lcd.getTextWidth(FONT_MINI,string.format("%.2fV",cellVolt)),16,string.format("%.2fV",cellVolt),FONT_MINI)
                dispLiPoClock(cellPerc,1,0)
    else -- Small window
        if (cellPerc==-1) then
            lcd.drawText(145-lcd.getTextWidth(FONT_BIG,"-%"),1,"-%",FONT_BIG)
            else
            lcd.drawText(145-lcd.getTextWidth(FONT_BIG,string.format("%s%%",cellPerc)),1,string.format("%s%%",cellPerc),FONT_BIG)
        end
        dispLiPoSmall(cellPerc,1,0)
    end
    collectgarbage()
end
----------------------------------------------------------------------
-- Store settings when changed by user
local function sensorChanged(value)
    sensId=sensorsAvailable[value].id
    sensPa=sensorsAvailable[value].param
    system.pSave("sensId",sensId)
    system.pSave("sensPa",sensPa)
end

local function cellCntChanged(value)
    cellCnt=value
    system.pSave("cellCnt",cellCnt)
end

local function alarmValChanged(value)
    alarmVal=value
    system.pSave("alarmVal",alarmVal)
end

local function alarmFileChanged(value)
    alarmFile=value
    system.pSave("alarmFile",alarmFile)
end
----------------------------------------------------------------------
-- Draw the main form (Application inteface)
local function initForm(subform)
    -- List sensors only if menu is active to preserve memory at runtime 
    -- (measured up to 25% save if menu is not opened)
    sensorsAvailable = {}
    local sensors = system.getSensors()
    local sensorTx = system.getTxTelemetry()
    local list={}
    local curIndex1 = -1
    local descr = ""
    -- Add some of RX Telemetry items to beginning in list of sensors, get name from translation
    list[#list + 1] = string.format("%s",trans19.sensorRx1)
    sensorsAvailable[#sensorsAvailable + 1] = {["unit"] = "V", ["param"] = 1,["id"] = 999,["label"] = trans19.sensorRx1}
    if(sensId == 999 and sensPa == 1) then
        curIndex1 = 1
    end
    list[#list + 1] = string.format("%s",trans19.sensorRx2)
    sensorsAvailable[#sensorsAvailable + 1] = {["unit"] = "V", ["param"] = 2,["id"] = 999,["label"] = trans19.sensorRx2}
    if(sensId == 999 and sensPa == 2) then
        curIndex1 = 2
    end
    list[#list + 1] = string.format("%s",trans19.sensorRxB)
    sensorsAvailable[#sensorsAvailable + 1] = {["unit"] = "V", ["param"] = 3,["id"] = 999,["label"] = trans19.sensorRxB}
    if(sensId == 999 and sensPa == 3) then
        curIndex1 = 3
    end
    -- Create a list of normal sensors
    for index,sensor in ipairs(sensors) do 
        if(sensor.param == 0) then
            descr = sensor.label
            else
            list[#list + 1] = string.format("%s - %s",descr,sensor.label)
            sensorsAvailable[#sensorsAvailable + 1] = sensor
            if(sensor.id == sensId and sensor.param == sensPa) then
                curIndex1 =# sensorsAvailable
            end
        end
    end
    collectgarbage()
    
    local form,addRow,addLabel=form,form.addRow,form.addLabel
    local addIntbox,addSelectbox=form.addIntbox,form.addSelectbox
    local addInputbox,addCheckbox=form.addInputbox,form.addCheckbox
    local addAudioFilebox,setButton=form.addAudioFilebox,form.setButton
    local addTextbox=form.addTextbox
    
    addRow(1)
    addLabel({label="---     RC-Thoughts Jeti Tools      ---",font=FONT_BIG})
    
    addRow(1)
    addLabel({label=trans19.labelSensor,font=FONT_BOLD})
    
    addRow(2)
    addLabel({label=trans19.sensorSel})
    --addSelectbox(sensList,curIndex,true,sensorChanged)
    addSelectbox(list,curIndex1,true,sensorChanged)
    
    addRow(2)
    addLabel({label=trans19.cellCount,width=220})
    addIntbox(cellCnt,0,24,0,0,1,cellCntChanged)
    
    addRow(1)
    addLabel({label=trans19.labelAlarm,font=FONT_BOLD})
    
    addRow(2)
    addLabel({label=trans19.alarmValue,width=220})
    addIntbox(alarmVal,0,99,0,0,1,alarmValChanged)
    
    form.addRow(2)
    addLabel({label=trans19.voiceFile})
    addAudioFilebox(alarmFile,alarmFileChanged)
    
    addRow(1)
    addLabel({label="Powered by RC-Thoughts.com-"..lipoVersion.." ",font=FONT_MINI,alignRight=true})
    
    formID=1
    collectgarbage()
end
----------------------------------------------------------------------
-- Table for binding cell-voltage to percentage
local percentList={{3,0},{3.093,1},{3.196,2},{3.301,3},{3.401,4},{3.477,5},{3.544,6},{3.601,7},{3.637,8},{3.664,9},{3.679,10},{3.683,11},{3.689,12},{3.692,13},{3.705,14},{3.71,15},{3.713,16},{3.715,17},{3.72,18},{3.731,19},{3.735,20},{3.744,21},{3.753,22},{3.756,23},{3.758,24},{3.762,25},{3.767,26},{3.774,27},{3.78,28},{3.783,29},{3.786,30},{3.789,31},{3.794,32},{3.797,33},{3.8,34},{3.802,35},{3.805,36},{3.808,37},{3.811,38},{3.815,39},{3.818,40},{3.822,41},{3.825,42},{3.829,43},{3.833,44},{3.836,45},{3.84,46},{3.843,47},{3.847,48},{3.85,49},{3.854,50},{3.857,51},{3.86,52},{3.863,53},{3.866,54},{3.87,55},{3.874,56},{3.879,57},{3.888,58},{3.893,59},{3.897,60},{3.902,61},{3.906,62},{3.911,63},{3.918,64},{3.923,65},{3.928,66},{3.939,67},{3.943,68},{3.949,69},{3.955,70},{3.961,71},{3.968,72},{3.974,73},{3.981,74},{3.987,75},{3.994,76},{4.001,77},{4.007,78},{4.014,79},{4.021,80},{4.029,81},{4.036,82},{4.044,83},{4.052,84},{4.062,85},{4.074,86},{4.085,87},{4.095,88},{4.105,89},{4.111,90},{4.116,91},{4.12,92},{4.125,93},{4.129,94},{4.135,95},{4.145,96},{4.176,97},{4.179,98},{4.193,99},{4.2,100}}
----------------------------------------------------------------------
-- Count percentage from cell voltage
function percCell(cellVoltage)
    result=0
    if(cellVoltage > 4.2 or cellVoltage < 3.00)then
        if(cellVoltage > 4.2)then
            result=100
        end
        if(cellVoltage < 3.00)then
            result=0
        end
        else
        for i,v in ipairs(percentList) do
            if(v[1] >= cellVoltage)then
                result=v[2]
                break
            end
        end
    end
    collectgarbage()
    return result
end
----------------------------------------------------------------------
-- Create a rolling average of voltage to prevent false alarms
function rollingAverage(period)
    local t={}
    function sum(a,...)
        if a then return a+sum(...) else return 0 end
    end
    function average(n)
        if #t==period then table.remove(t,1) end
        t[#t+1]=n
        return sum(table.unpack(t)) / #t
    end
    collectgarbage()
    return average
end
----------------------------------------------------------------------
-- Runtime functions,read sensor,convert to percentage
local function loop()
    local sensor = {}
    local sensorRx = system.getTxTelemetry()
    if(sensId == 999) then
        if(sensPa == 1) then
            sensor.value = sensorRx.rx1Voltage
            elseif (sensPa == 2) then
            sensor.value = sensorRx.rx2Voltage
            elseif (sensPa == 3) then
            sensor.value = sensorRx.rxBVoltage
        end
        if(sensorRx.rx1Percent > 1)then
            sensor.valid = true
        else
            sensor.valid = false
        end
        else
        sensor = system.getSensorByID(sensId,sensPa)
    end

    local timeNow = system.getTimeCounter()

    if(sensor and sensor.valid) then
        voltTot=sensor.value
        -- Fill table at start
        if(timeFill == 0) then
            timeFill = timeNow + 2000
        end
        if(timeNow <= timeFill) then
            cellVolt = (roll(voltTot)/cellCnt)
        end
        -- 
        if(timeNow >= timeLast + 1000) then
            cellVolt = (roll(voltTot)/cellCnt)
            timeLast = timeNow
            voltMax = 0
        end
        -- Track maximum voltage seen during this collection cycle
		if voltTot > voltMax then	
			voltMax = voltTot
		end
        
        -- Calculate cell-percentage from LiPo-table
        cellPerc=percCell(cellVolt)
        -- Take care of alarm if alarm is enabled, make sure it plays only once
        if(not playDone and alarmVal > 0 and cellPerc <= alarmVal and alarmFile ~= "...") then
            system.playFile(alarmFile,AUDIO_QUEUE)
            system.playNumber(cellPerc,0,"%")
            playDone=true   
        end
        else
        -- If we have no sensor set cell-percentage to -1 for screen-identification
        cellPerc = -1
        timeFill = 0
    end   
    -- If percentage is above alarm-level enable alarm
    if(cellPerc > alarmVal) then
        playDone=false
    end
    collectgarbage()
end
----------------------------------------------------------------------
-- Application initialization
local function init()
    local pLoad,registerForm,registerTelemetry=system.pLoad,system.registerForm,system.registerTelemetry
    sensId=pLoad("sensId",0)
    sensPa=pLoad("sensPa",0)
    cellCnt=pLoad("cellCnt",0)
    alarmVal=pLoad("alarmVal",0)
    alarmFile=pLoad("alarmFile","...")
    registerForm(1,MENU_APPS,trans19.appName,initForm)
    registerTelemetry(1,trans19.appName,0,dispLiPo)
    -- Set average-calculation
    roll = rollingAverage(10)
    collectgarbage()
end
----------------------------------------------------------------------
lipoVersion="v.1.3"
setLanguage()
collectgarbage()
return {init=init,loop=loop,author="RC-Thoughts",version=lipoVersion,name=trans19.appName}
